# Анализ базы сервиса чтения книг.

**Задача**: проанализировать базу данных крупного сервиса чтения книг по подписке, данные будут использованы для ценностного предложения для нового продукта.

Анализ производится на основании данных, предоставленных заказчиком.

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Содержание.
1. [Загрузка данных.](#start)
    * [Вывод](#resume)
2. [Книги, выпущенные после 01.01.2020](#books_after)
3. [Количество обзоров и средняя оценка книг.](#raitings) 
4. [Издательство, которое выпустило наибольшее число книг толще 50 страниц.](#big_books)
5. [Автор, с самой высокой средней оценкой книг.](#big_author)
6. [Обзоры пользователей.](#reviews)
7. [Вывод.](#resume1)

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

<a id='start'></a>

## Загрузка данных.

In [2]:
# устанавливаем параметры
db_config = {'user': '', # имя пользователя
             'pwd': '', # пароль
             'host': '',
             'port': , # порт подключения
             'db': ''} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [13]:
query = ''' SELECT
                book_id,
                author_id,
                title,
                num_pages,
                publication_date,
                publisher_id
            FROM
                books
        '''
books = pd.io.sql.read_sql(query, con=engine)

pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

В базе данных сервиса 1000 книг.

In [4]:
query = ''' SELECT
                author_id,
                author
            FROM
                authors
        '''
authors = pd.io.sql.read_sql(query, con=engine)

pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

В базе данных сервиса 636 авторов.

In [5]:
query = ''' SELECT
                publisher_id,
                publisher
            FROM
                publishers
        '''
publishers = pd.io.sql.read_sql(query, con=engine)

pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

В базе данных сервиса 340 издательств.

In [6]:
query = ''' SELECT
                rating_id,
                book_id,
                username,
                rating
            FROM
                ratings
        '''
ratings = pd.io.sql.read_sql(query, con=engine, index_col='rating_id')

pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

В базе данных сервиса 6456 пользовательских оценок книг.

In [7]:
query = ''' SELECT
                review_id,
                book_id,
                username,
                text
            FROM
                reviews
        '''
reviews = pd.io.sql.read_sql(query, con=engine)

pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

В базе данных сервиса 2793 пользовательских обзоров на книги.

<a id='resume'></a>

## Вывод.
В базе данных сервиса есть 1000 книг 636 авторов, выпущенных 340 издательствами. На книги есть 2793 пользовательских обзора и 6456 пользовательских оценок.

<a id='books_after'></a>

## Книги, выпущенные после 01.01.2020.

Посчитать, сколько книг вышло после 1 января 2000 года.

In [8]:
query = '''SELECT
               COUNT(publication_date) AS books_cnt
           FROM
               books 
           WHERE
               publication_date > '01-01-2000';
        '''
books_cnt = pd.io.sql.read_sql(query, con=engine)
pd.io.sql.read_sql(query, con = engine)

books_cnt
0        819

В базе данных сервиса 819 книг, выпущенных после 01.01.2000

<a id='raitings'></a>

## Количество обзоров и средняя оценка.

Для каждой книги посчитать количество обзоров и среднюю оценку.

In [9]:
query = '''SELECT
               books.book_id AS book_id,
               books.title AS title,
               AVG(ratings.rating) AS mean_rating,
               COUNT(reviews.review_id) AS review_count
           FROM
               books
               INNER JOIN ratings ON ratings.book_id = books.book_id
               INNER JOIN reviews ON reviews.book_id = books.book_id
           GROUP BY
               books.book_id,
               books.title
           ORDER BY
               mean_rating DESC,
               review_count DESC;    
        '''
books_rating = pd.io.sql.read_sql(query, con=engine)
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  mean_rating  \
0         17                      A Dirty Job (Grim Reaper  #1)         5.00   
1        553            School's Out—Forever (Maximum Ride  #2)         5.00   
2        444       Moneyball: The Art of Winning an Unfair Game         5.00   
3        347  In the Hand of the Goddess (Song of the Liones...         5.00   
4        229           Evil Under the Sun (Hercule Poirot  #24)         5.00   
..       ...                                                ...          ...   
989      915  The World Is Flat: A Brief History of the Twen...         2.25   
990      202                                      Drowning Ruth         2.00   
991      371                                              Junky         2.00   
992      316                  His Excellency: George Washington         2.00   
993      303                               Harvesting the Heart         1.50   

     review_count  
0              16  
1              12  
2               9  
3               6  
4               4  
..            ...  
989            12  
990             9  
991             4  
992             4  
993             4  

[994 rows x 4 columns]

Средняя оценка книг составляет от 5 до 1.5 баллов.

<a id='big_books'></a>

## Издательство, которое выпустило наибольшее число книг толще 50 страниц.

Определить издательство, которое выпустило наибольшее число книг толще 50 страниц.

In [14]:
query = '''SELECT
               publishers.publisher_id AS publisher_id,
               publishers.publisher AS publisher,
               COUNT(DISTINCT books.book_id) AS books_count
           FROM
               publishers
               INNER JOIN books ON publishers.publisher_id = books.publisher_id
           WHERE
               books.num_pages > 50
           GROUP BY
               publishers.publisher_id,
               publishers.publisher
           ORDER BY
               books_count DESC
           LIMIT 1;
        '''
publishers_books = pd.io.sql.read_sql(query, con=engine)
pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher  books_count
0           212  Penguin Books           42

Издательство Penguin Books выпустило максимальное количество книг толще 50 страниц из базы сервиса - 42.

<a id='big_author'></a>

## Автор, с самой высокой средней оценкой книг.

Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками.

Исходя из формулировки запроса считаю, что в данном случае нужно определить среднюю оценку каждой книги, а потом определить среднюю оценку книг по каждому автору.

In [11]:
query = '''SELECT
               authors.author_id AS author_id,
               authors.author AS author,
               AVG(Sub.rating_mean) AS mean_rating
           FROM (
               SELECT
                   books.book_id AS book_id,
                   books.author_id AS author_id,
                   COUNT(DISTINCT ratings.rating_id) AS ratings_count,
                   AVG(ratings.rating) AS rating_mean
               FROM
                   ratings
                   INNER JOIN books ON ratings.book_id = books.book_id
               GROUP BY
                   books.book_id,
                   books.author_id
               HAVING
                   COUNT(DISTINCT ratings.rating_id) > 49) AS Sub
               LEFT JOIN authors ON Sub.author_id = authors.author_id
           GROUP BY
               authors.author_id,
               authors.author
           ORDER BY
               mean_rating DESC
           LIMIT 1;
        '''
big_author = pd.io.sql.read_sql(query, con=engine)
pd.io.sql.read_sql(query, con = engine)

author_id                      author  mean_rating
0        236  J.K. Rowling/Mary GrandPré     4.283844

Автор книг с самой высокой средней оценкой - Дж.К. Роулинг, иллюстратор этих книг - Мэри ГранПре, скорее всего это американские из дания книг о Гарри Поттере.

<a id='reviews'></a>

## Обзоры пользователей.

Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [12]:
query = '''SELECT
               username,
               AVG(count_reviews) AS mean_cnt_review
           FROM (
               SELECT
                   SubQ.username AS username,
                   COUNT(reviews.review_id) AS count_reviews
               FROM (
                   SELECT
                       username,
                       COUNT(rating_id) AS rating_count
                   FROM
                       ratings
                   GROUP BY
                       username
                   HAVING
                       COUNT(rating_id) > 50) AS SubQ
               LEFT JOIN reviews ON reviews.username = SubQ.username
               GROUP BY
                   SubQ.username) AS SubQ1
           GROUP BY
               username
           ORDER BY
               mean_cnt_review DESC;
        '''
users_reviews = pd.io.sql.read_sql(query, con=engine)
pd.io.sql.read_sql(query, con = engine)

username  mean_cnt_review
0     sfitzgerald             28.0
1      martinadam             27.0
2       richard89             26.0
3  jennifermiller             25.0
4          paul88             22.0
5          xdavis             18.0

Пользователей, которые поставили больше 50 оценок всего 6, при этом эти же активные пользователи пишут большое количество отзывов на книги.

<a id='resume1'></a>

## Вывод.

В базе данных сервиса есть 1000 книг 636 авторов, выпущенных 340 издательствами. 

Большую часть фонда составляют книги, выпущенные после 01.01.2020 - 819 книг. 

Издательство Penguin Books выпустило максимальное количество книг толще 50 страниц из базы сервиса - 42 книги.

На книги есть 2793 пользовательских обзора и 6456 пользовательских оценок. Средняя оценка книг составляет от 5 до 1.5 баллов.

Пользователей, которые поставили больше 50 оценок всего 6, при этом эти же активные пользователи пишут большое количество отзывов на книги.

Автор книг с самой высокой средней оценкой - Дж.К. Роулинг, иллюстратор этих книг - Мэри ГранПре, скорее всего это американские издания книг о Гарри Поттере.